src/main/java/org/takes/facets/hamcrest/AbstractHmTextBody.java

```diff
@@ -0,0 +1,119 @@
+/**
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2014-2018 Yegor Bugayenko
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package org.takes.facets.hamcrest;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.Charset;
+import org.hamcrest.Description;
+import org.hamcrest.Matcher;
+import org.hamcrest.TypeSafeMatcher;
+
+/**
+ * Text body matcher.
+ *
+ * <p>This "matcher" tests given item body, assuming that it has text content.
+ * <p>The class is immutable and thread-safe.
+ *
+ * @author Izbassar Tolegen (t.izbassar@gmail.com)
+ * @version $Id$
+ * @param <T> Item type. Should be able to return own body
+ * @since 2.0
+ *
+ * @todo #794:30min Implement describeMismatchSafely and cover mismatch
+ *  descriptions with relevant test cases. It should show, what was
+ *  expected and what was actually in the body for clear understanding.
+ */
+public abstract class AbstractHmTextBody<T> extends TypeSafeMatcher<T> {
+
+    /**
+     * Body matcher.
+     */
+    private final Matcher<String> body;
+
+    /**
+     * Charset of the text.
+     */
+    private final Charset charset;
+
+    /**
+     * Ctor.
+     * @param body Body matcher.
+     * @param charset Charset of the text.
+     */
+    public AbstractHmTextBody(final Matcher<String> body,
+        final Charset charset) {
+        super();
+        this.body = body;
+        this.charset = charset;
+    }
+
+    @Override
+    public final void describeTo(final Description description) {
+        description.appendDescriptionOf(this.body);
```

Comment:
@t-izbassar it will be not really informative, it should include not only expected result, but also actual value and some text description. Please see how it is implemented in `AbstractHmHeader`

Reply:
@g4s8 updated puzzle description to cover that issue.